sf-stream/src/runtime.rs

```diff
@@ -0,0 +1,227 @@
+                            "[sf-stream] oldest txn version is {} but requested version is {}",
+                            oldest_version, &self.current_version
+                        );
+                        sleep(Duration::from_millis(300)).await;
+                        return result;
+                    }
+                }
+            }
+            Err(err) => {
+                warn!("[sf-stream] failed to get first txn version: {}", err);
+                sleep(Duration::from_millis(300)).await;
+                return result;
+            }
+        }
+
+        let ledger_info = self.context.get_latest_ledger_info().unwrap();
+        match self
+            .context
+            .get_transactions(self.current_version, batch_size, ledger_info.version())
+        {
+            Ok(transactions) => {
+                if transactions.is_empty() {
+                    debug!("[sf-stream] no transactions to send");
+                    sleep(Duration::from_millis(100)).await;
+                    return result;
+                }
+                // TODO: there might be an off by one (tx version fetched)
+                debug!(
+                    "[sf-stream] got {} transactions from {} to {} [{}]",
+                    transactions.len(),
+                    self.current_version,
+                    self.current_version + transactions.len() as u64,
+                    transactions.last().map(|txn| txn.version).unwrap_or(0)
+                );
+                for onchain_txn in transactions {
+                    // TODO: assert txn.version == &self.current_version + 1?
+                    // TODO: return a `Result` & check to ensure we pushed before incrementing
+                    let txn_version = onchain_txn.version;
+                    // Todo: since the timestamp is per block, only calculate this value once per block
+                    let timestamp = self
+                        .context
+                        .get_block_timestamp(onchain_txn.version)
+                        .unwrap_or_else(|_| {
+                            panic!(
+                                "Could not get timestamp for version {}",
+                                onchain_txn.version
+                            )
+                        });
+                    let txn = self
+                        .resolver
+                        .as_converter(self.context.db.clone())
+                        .try_into_onchain_transaction(timestamp, onchain_txn)
+                        .unwrap_or_else(|e| {
+                            panic!(
+                                "Could not convert onchain transaction version {} into transaction: {:?}",
+                                txn_version, e
+                            )
+                        });
+
+                    let txn_proto = self.convert_transaction(txn);
+                    self.print_transaction(&txn_proto).await;
+                    result.push(txn_proto);
```

Comment:
Why do you collect these if you're just printing them out?  Should you print the transactions in a batch?